src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -24,6 +24,45 @@ License
 \*---------------------------------------------------------------------------*/
 
 #include "directAMI.H"
+#include "primitiveFieldsFwd.H"
+#include "pointFieldFwd.H"
+#include "faceData.H"
+
+namespace Foam {
+    struct ShootRayFunctor : public std::binary_function<label, label, void> {
+        const pointField *tgtPoints;
+        const pointField *srcPoints;
+        const vectorField *srcCf;
+        const faceData *tgtFaces;
+        const faceData *srcFaces;
+
+        ShootRayFunctor
+        (
+        const pointField *_tgtPoints,
+        const pointField *_srcPoints,
+        const vectorField *_srcCf,
+        const faceData *_tgtFaces,
+        const faceData *_srcFaces
+        ) :
+        tgtPoints(_tgtPoints),
+        srcPoints(_srcPoints),
+        srcCf(_srcCf),
+        tgtFaces(_tgtFaces),
+        srcFaces(_srcFaces)
+        {}
+
+        __HOST____DEVICE__
+        void operator()(const label& tgtI, const label& srcI) {
+            pointHit ray = tgtFaces[tgtI].ray(srcCf[srcI], srcFaces[srcI].normal(srcPoints), tgtPoints);
+            if (ray.hit()) {
+                // found an intersection !
+                // Doesn't work currently since the "ray" method is
+                // commented out in faceData.H
+                // We should update the addressing here.
```

Comment:
I attempted to port the `ray` method to the GPU [here](https://github.com/Atizar/RapidCFD-dev/pull/7/files#diff-1e5365ff76d190175d70b2a0e0074827R162).
Please have a look.